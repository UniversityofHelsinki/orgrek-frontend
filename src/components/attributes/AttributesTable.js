import React from 'react';
import Table from '@mui/material/Table';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import TableCell from '@mui/material/TableCell';
import TableBody from '@mui/material/TableBody';
import TableContainer from '@mui/material/TableContainer';
import Validity from './Validity';
import { useTranslation } from 'react-i18next';

const AttributesTable = ({ columns, data, summary, keyFn, sx }) => {
  const { t } = useTranslation();

  const defaultColumns = [
    { label: t('attribute.label'), render: (item) => t(item.key) },
    { label: t('value'), render: (item) => t(item.value) },
    {
      label: t('valid_dates'),
      render: (item) => (
        <Validity startDate={item.startDate} endDate={item.endDate} />
      ),
    },
  ];
  const defaultKeyFn = (item) =>
    `${item.key}-${item.value}-${item.startDate}-${item.endDate}`;

  const cols = columns || defaultColumns;
  const getKey = keyFn || defaultKeyFn;

  const renderHeader = (column) => (
    <TableCell key={column.label}>{column.label}</TableCell>
  );

  const renderCell = (column, item) => (
    <TableCell key={column.label}>{column.render(item)}</TableCell>
  );

  const renderedHeaders = cols.map((column) => {
    if (column.renderHeader) {
      return column.renderHeader(column);
    } else {
      return renderHeader(column);
    }
  });

  const renderedRows = data.map((item, index) => {
    const renderedCells = cols.map((column) => {
      if (column.renderCell) {
        return column.renderCell(column, item);
      } else {
        return renderCell(column, item);
      }
    });

    return <TableRow key={getKey(item, index)}>{renderedCells}</TableRow>;
  });

  return (
    <TableContainer
      component="div"
      data-testid="attributesTable"
      sx={[{ mb: 2 }, ...(Array.isArray(sx) ? sx : [sx])]}
    >
      <Table size="small" aria-label={summary} summary={summary}>
        <TableHead>
          <TableRow>{renderedHeaders}</TableRow>
        </TableHead>
        <TableBody>{renderedRows}</TableBody>
      </Table>
    </TableContainer>
  );
};

export default AttributesTable;
