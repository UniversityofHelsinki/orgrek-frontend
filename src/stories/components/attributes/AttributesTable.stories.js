import AttributesTable from '../../../components/attributes/AttributesTable';

export default {
  component: AttributesTable,
};

export const Default = {
  args: {
    summary: 'Nimitiedot',
    data: [
      {
        id: 4821,
        nodeId: '4820',
        key: 'name_en',
        value: 'IT Solutions',
        startDate: null,
        endDate: null,
      },
      {
        id: 4822,
        nodeId: '4820',
        key: 'name_sv',
        value: 'Datateknikl√∂sningar',
        startDate: null,
        endDate: null,
      },
      {
        id: 4826,
        nodeId: '4820',
        key: 'name_fi',
        value: 'Tietotekniikkaratkaisut',
        startDate: null,
        endDate: null,
      },
    ],
  },
};
