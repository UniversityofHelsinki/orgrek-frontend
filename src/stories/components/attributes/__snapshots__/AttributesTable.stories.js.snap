// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/attributes/AttributesTable Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Attribuutti",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Arvo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Englanniksi",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "IT Solutions",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Ruotsiksi",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Datatekniklösningar",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Suomeksi",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Tietotekniikkaratkaisut",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-attributes-attributestable--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/attributes/AttributesTable DuplicateKey smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Attribuutti",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Arvo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaolo",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Lyhenne",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "value1",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Lyhenne",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "value2",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Voimassaoloa ei määritetty",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-attributes-attributestable--duplicate-key",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
